modules/consensus/synchronize.go

```diff
@@ -11,54 +11,76 @@ import (
 )
 
 const (
-	MaxCatchUpBlocks       = 10
-	MaxSynchronizeAttempts = 8
+	// MaxCatchUpBlocks is the maxiumum number of blocks that can be given to
+	// the consensus set in a single iteration during the initial blockchain
+	// download.
+	MaxCatchUpBlocks = 10
 )
 
-// blockHistory returns up to 32 BlockIDs, starting with the 12 most recent
-// BlockIDs and then doubling in step size until the genesis block is reached.
-// The genesis block is always included. This array of BlockIDs is used to
-// establish a shared commonality between peers during synchronization.
-func (s *ConsensusSet) blockHistory() (blockIDs [32]types.BlockID) {
-	knownBlocks := make([]types.BlockID, 0, 32)
+// blockHistory returns up to 32 block ids, starting with recent blocks and
+// then proving exponentially increasingly less recent blocks. The genesis
+// block is always included as the last block. This block history can be used
+// to find a common parent that is reasonably recent, usually the most recent
+// common parent is found, but always a common parent within a factor of 2 is
+// found.
+func blockHistory(tx *bolt.Tx) (blockIDs [32]types.BlockID) {
+	height := blockHeight(tx)
 	step := types.BlockHeight(1)
-	for height := s.height(); ; height -= step {
-		// after 12, start doubling
-		knownBlocks = append(knownBlocks, s.db.getPath(height))
-		if len(knownBlocks) >= 12 {
+	// The final step is to include the genesis block, which is why `step <
+	// height` is used as opposed to `step <= height`.
```

Comment:
I realized that I was confused about how this function works. You probably don't need to double all the way out to the genesis block, and are unlikely to ever farther than the 3rd element unless you've been victim to an isolation attack.